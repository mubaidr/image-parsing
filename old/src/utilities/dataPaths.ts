import os from 'os'
import path from 'path'

const isDev: boolean = process.env.NODE_ENV === 'development'
const tmp: string = isDev ? path.resolve('.tmp') : os.tmpdir()
const home: string = path.join(os.homedir(), 'desktop')

const dataPaths = {
  design: path.resolve('__tests__', '_data', 'design.qr.svg'),
  designBarcode: path.resolve('__tests__', '_data', 'design.svg'),
  home,
  images: path.resolve('__tests__', '_data', 'images-qrcode'),
  imagesBarcode: path.resolve('__tests__', '_data', 'images-barcode'),
  key: path.resolve('__tests__', '_data', 'key.xlsx'),
  keyImage: path.resolve('__tests__', '_data', 'key.jpg'),
  questionsModel: isDev
    ? path.resolve('src/data/questions-model.json')
    : path.resolve('dist/data/questions-model.json'),
  result: path.resolve('__tests__', '_data', 'result.xlsx'),
  resultCompiled: path.resolve('__tests__', '_data', 'resultCompiled.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
}

export { dataPaths }
