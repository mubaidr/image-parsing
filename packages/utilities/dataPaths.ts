import os from 'os';
import path from 'path';

const tmp = os.tmpdir();

export const DataPaths = {
  design: path.resolve('tests', '_data', 'design.qr.svg'),
  designBarcode: path.resolve('tests', '_data', 'design.svg'),
  home: os.homedir(),
  images: path.resolve('tests', '_data', 'images-qrcode'),
  imagesBarcode: path.resolve('tests', '_data', 'images-barcode'),
  key: path.resolve('tests', '_data', 'key.xlsx'),
  keyImage: path.resolve('tests', '_data', 'key.jpg'),
  result: path.resolve('tests', '_data', 'result.xlsx'),
  resultCompiled: path.resolve('tests', '_data', 'resultCompiled.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
};
