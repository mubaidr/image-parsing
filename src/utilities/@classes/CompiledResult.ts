import uuid from 'uuid'

import { importExcelToJson } from '../excel'
import Result from './Result'

class CompiledResult {
  private id: string
  private lastSavedTime: Date | undefined

  private keys: Result[] = []
  private results: Result[] = []

  public constructor() {
    this.id = uuid()
  }

  public getKeyCount(): number {
    return this.keys.length
  }

  public getResultCount(): number {
    return this.results.length
  }

  public getKeys(): Result[] {
    return this.keys
  }

  public getResults(): Result[] {
    return this.results
  }

  public addKeys(key: Result | Result[]): CompiledResult {
    if (key instanceof Result) {
      this.keys.push(key)
    } else {
      this.keys.push(...key)
    }

    return this
  }

  public addResults(result: Result | Result[]): CompiledResult {
    if (result instanceof Result) {
      this.results.push(result)
    } else {
      this.results.push(...result)
    }

    return this
  }

  public compile(marks?: number, negativeMarks?: number): CompiledResult {
    if (this.keys.length === 0) throw 'Keys not loaded'
    if (this.results.length === 0) throw 'Results not loaded'

    for (let i = 0; i < this.keys.length; i += 1) {
      for (let j = 0; j < this.results.length; j += 1) {
        this.results[j].compile(this.keys[i], marks, negativeMarks)
      }
    }

    return this
  }

  public static loadFromExcel(src: string): CompiledResult {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const rows: any[] = importExcelToJson(src)
    const compiledResult = new CompiledResult()

    rows.forEach(row => {
      const result = Result.fromJson(row)

      if (result.isKey()) {
        compiledResult.addKeys(result)
      } else {
        compiledResult.addResults(result)
      }
    })

    return compiledResult
  }

  public addFromExcel(src: string): CompiledResult {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const rows: any[] = importExcelToJson(src)

    rows.forEach(row => {
      const result = Result.fromJson(row)

      if (result.isKey()) {
        this.addKeys(result)
      } else {
        this.addResults(result)
      }
    })

    return this
  }

  public static merge(compiledResults: CompiledResult[]): CompiledResult {
    const compiledResult = new CompiledResult()

    compiledResults.forEach(cr => {
      compiledResult.addKeys(cr.getKeys()).addResults(cr.getResults())
    })

    return compiledResult
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public export(): any[] {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const obj: any[] = []

    this.results.forEach(result => {
      obj.push(result.toJson())
    })

    return obj
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public static export(compiledResult: CompiledResult): any[] {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const obj: any[] = []

    compiledResult.results.forEach(result => {
      obj.push(result.toJson())
    })

    return obj
  }
}

export default CompiledResult
