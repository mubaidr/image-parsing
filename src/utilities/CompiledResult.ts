import { v4 as uuidv4 } from 'uuid'
import { importExcelToJson } from './excel'
import { Result } from './Result'

export class CompiledResult {
  id: string
  lastSavedTime: Date | undefined
  keys: Result[] = []
  results: Result[] = []

  constructor() {
    this.id = uuidv4()
  }

  static loadFromExcel(src: string): CompiledResult {
    const rows: unknown[] = importExcelToJson(src)
    const compiledResult = new CompiledResult()

    rows.forEach((row) => {
      const result = Result.fromJson(row)

      if (result.isKey()) {
        compiledResult.addKeys([result])
      } else {
        compiledResult.addResults([result])
      }
    })

    return compiledResult
  }

  addKeys(keys: Result[]): CompiledResult {
    this.keys.push(...keys)
    return this
  }

  addResults(results: Result[]): CompiledResult {
    this.results.push(...results)
    return this
  }

  getKeysAndResults(): Result[] {
    return [...this.keys, ...this.results]
  }

  compile(marks?: number, negativeMarks?: number): CompiledResult {
    if (this.keys.length === 0) throw new Error('Keys not loaded')
    if (this.results.length === 0) throw new Error('Results not loaded')

    for (let i = 0; i < this.keys.length; i += 1) {
      for (let j = 0; j < this.results.length; j += 1) {
        this.results[j].compile(this.keys[i], marks, negativeMarks)
      }
    }

    return this
  }

  sortResults(): CompiledResult {
    this.results.sort((a, b) => {
      return parseInt(a.rollNo || '0', 10) - parseInt(b.rollNo || '0', 10)
    })

    return this
  }

  static merge(compiledResults: CompiledResult[]): CompiledResult {
    const compiledResult = new CompiledResult()

    compiledResults.forEach((cr) => {
      compiledResult.addKeys(cr.keys).addResults(cr.results)
    })

    return compiledResult
  }
}
