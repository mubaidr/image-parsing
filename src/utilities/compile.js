const { NATIVE_KEYS } = require('../utilities/valid-types')
const { importExcelToJSON } = require('../utilities/excel')
const { getDesignData } = require('./design')
const { processTask } = require('./processTaskWorker')
const dataPaths = require('./data-paths')

async function readResult(src) {
  return importExcelToJSON(src)
}

async function readKey(src) {
  const isNativeKey = NATIVE_KEYS.indexOf(src.split('.').pop()) !== -1

  if (isNativeKey) {
    return importExcelToJSON(src)
  }

  const designData = await getDesignData(dataPaths.DEFAULTS.design)
  const keyData = await processTask(designData, [src])

  return keyData
}

async function compileResult(resultPath, keyPath, options) {
  const compiledResult = []
  const [results, keys] = await Promise.all([
    readResult(resultPath),
    readKey(keyPath),
  ])

  // TODO support multiple keys in one file
  const key = keys[0]

  for (let i = 0; i < results.length; i += 1) {
    const result = results[i]

    // skip processed result
    if (result.isProcessed) continue

    // compare with key
    let totalQuestions = 0
    let skippedQuestions = 0
    let skippedAnswers = 0
    let attemptedAnswers = 0
    let correctAnswers = 0
    let incorrectAnswers = 0

    const columns = Object.keys(key).filter(col => col[0].toLowerCase() === 'q')

    // total questions
    totalQuestions = columns.length

    for (let j = 0; j < columns.length; j += 1) {
      const column = columns[j]

      // question is skipped in key
      if (
        !key[column] ||
        ['?', '*', '', ' ', undefined, null, false, 0].indexOf(key[column]) >= 0
      ) {
        skippedQuestions += 1
        continue
      }

      // question is skipped by candidate
      if (result[column] === '?') {
        skippedAnswers += 1
        continue
      }

      // other questions are attempted
      attemptedAnswers += 1

      // check correct/incorrect options
      if (result[column] === key[column]) {
        correctAnswers += 1
      } else {
        incorrectAnswers += 1
      }
    }

    compiledResult.push({
      ...result,
      totalQuestions,
      skippedQuestions,
      skippedAnswers,
      attemptedAnswers,
      correctAnswers,
      incorrectAnswers,
      totalMarks: totalQuestions * options.correctMarks,
      obtainedMarks:
        correctAnswers * options.correctMarks -
        incorrectAnswers * options.incorrectMarks,
    })

    // update status of result to skip it for next iterations
    result.isProcessed = true
  }

  return compiledResult
}

module.exports = {
  compileResult,
}
