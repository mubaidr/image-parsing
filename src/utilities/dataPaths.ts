import os from 'os'
import path from 'path'

const isDev: boolean = process.env.NODE_ENV === 'development'

const tmp: string = isDev ? path.resolve('.tmp') : os.tmpdir()
const home: string = path.join(os.homedir(), 'desktop')

const dataPaths = {
  design: path.resolve('_test_data', 'design.qr.svg'),
  designBarcode: path.resolve('_test_data', 'design.svg'),
  home,
  images: path.resolve('_test_data', 'images-qrcode'),
  imagesBarcode: path.resolve('_test_data', 'images-barcode'),
  key: path.resolve('_test_data', 'key.xlsx'),
  keyImage: path.resolve('_test_data', 'key.jpg'),
  questionsModel: path.resolve('src/data/questions-model.json'),
  result: path.resolve('_test_data', 'result.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
}

export { dataPaths }
