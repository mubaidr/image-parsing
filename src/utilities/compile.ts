import { VALIDTYPES } from '../utilities/validTypes'
import { DATAPATHS } from './dataPaths'

import IKey from '../@interfaces/IKey'

import { importExcelToJson } from '../utilities/excel'
import { getDesignData } from './design'
import { processTask } from './processTaskWorker'

type ReadKeyGetter = (a: string) => Promise<IKey[]>

const readKey: ReadKeyGetter = async src => {
  const ext = src.split('.').pop()

  if (ext === undefined) {
    return false
  }

  if (VALIDTYPES.NATIVE_KEYS.indexOf(ext) !== -1) {
    return importExcelToJson(src)
  }

  const designData = await getDesignData(DATAPATHS.design)
  return processTask(designData, [src])
}

type CompileResultGetter = (a: string, b: string) => Promise<object>

const compileResult: CompileResultGetter = async (resultPath, keyPath) => {
  const [results, keys] = await Promise.all([
    importExcelToJson(resultPath),
    readKey(keyPath),
  ])

  const key = keys[0]
  const compiledResults = []
  const IS_PROCESSED: string[] = []

  for (const result of results) {
    // skip processed result
    if (IS_PROCESSED.indexOf(result.ROLL_NO)) {
      continue
    }

    const columns = Object.keys(key).filter(col => col[0].toLowerCase() === 'q')
    const compiledResult = {
      attemptedAnswers: 0,
      correctAnswers: 0,
      incorrectAnswers: 0,
      skippedAnswers: 0,
      skippedQuestions: 0,
      totalQuestions: 0,
    }

    for (const column of columns) {
      // question is skipped in key
      if (
        !key[column] ||
        ['?', '*', '', ' ', undefined, null, false, 0].indexOf(key[column]) >= 0
      ) {
        compiledResult.skippedQuestions += 1
        continue
      }

      // question is skipped by candidate
      if (result[column] === '?') {
        compiledResult.skippedAnswers += 1
        continue
      }

      // check correct/incorrect options
      if (result[column] === key[column]) {
        compiledResult.correctAnswers += 1
      } else {
        compiledResult.incorrectAnswers += 1
      }

      // other questions are attempted
      compiledResult.attemptedAnswers += 1
    }

    // total questions
    compiledResult.totalQuestions = columns.length

    // collect result
    compiledResults.push({ ...result, ...compiledResult })

    // update status of result to skip it for next iterations
    IS_PROCESSED.push(result.ROLL_NO)
  }

  return compiledResults
}

module.exports = {
  compileResult,
}
