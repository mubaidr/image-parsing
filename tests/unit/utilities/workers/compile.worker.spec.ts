import { CompiledResult } from '@/utilities/CompiledResult'
import { dataPaths } from '@/utilities/dataPaths'
import { readKey } from '@/utilities/readKey'
import { start } from '@/utilities/workers/compile.worker'

describe('workerCompile', () => {
  test('should compile just fine', async () => {
    const results = CompiledResult.loadFromExcel(dataPaths.result).getResults()
    const keys = await readKey(dataPaths.key)

    if (!keys) fail()

    const compiledResult = await start(
      {
        results,
        keys,
      },
      false
    )

    expect(compiledResult).toBeDefined()
    expect(compiledResult?.getKeyCount()).toBeGreaterThanOrEqual(1)
    expect(compiledResult?.getResultCount()).toBeGreaterThanOrEqual(3)
  })
})
