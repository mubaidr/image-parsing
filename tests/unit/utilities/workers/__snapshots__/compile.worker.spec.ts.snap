// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workerCompile should compile just fine 1`] = `
Object {
  "answers": Object {
    "q1": Object {
      "correct": true,
      "value": "d",
    },
    "q10": Object {
      "correct": false,
      "value": "d",
    },
    "q11": Object {
      "correct": false,
      "value": "c",
    },
    "q12": Object {
      "correct": true,
      "value": "d",
    },
    "q13": Object {
      "correct": false,
      "value": "c",
    },
    "q14": Object {
      "correct": false,
      "value": "b",
    },
    "q15": Object {
      "correct": false,
      "value": "a",
    },
    "q16": Object {
      "correct": false,
      "value": "b",
    },
    "q17": Object {
      "correct": false,
      "value": "c",
    },
    "q18": Object {
      "correct": false,
      "value": "b",
    },
    "q19": Object {
      "correct": true,
      "value": "c",
    },
    "q2": Object {
      "correct": false,
      "value": "d",
    },
    "q20": Object {
      "correct": false,
      "value": "a",
    },
    "q21": Object {
      "correct": false,
      "value": "d",
    },
    "q22": Object {
      "correct": false,
      "value": "d",
    },
    "q23": Object {
      "correct": false,
      "unattempted": true,
      "value": "?",
    },
    "q24": Object {
      "correct": false,
      "value": "d",
    },
    "q25": Object {
      "correct": false,
      "value": "d",
    },
    "q26": Object {
      "correct": false,
      "value": "c",
    },
    "q27": Object {
      "correct": false,
      "value": "c",
    },
    "q28": Object {
      "correct": false,
      "value": "a",
    },
    "q29": Object {
      "correct": false,
      "value": "b",
    },
    "q3": Object {
      "correct": false,
      "value": "c",
    },
    "q30": Object {
      "correct": false,
      "value": "c",
    },
    "q31": Object {
      "correct": true,
      "value": "b",
    },
    "q32": Object {
      "correct": true,
      "value": "a",
    },
    "q33": Object {
      "correct": false,
      "value": "d",
    },
    "q34": Object {
      "correct": true,
      "value": "b",
    },
    "q35": Object {
      "correct": true,
      "value": "a",
    },
    "q36": Object {
      "correct": false,
      "value": "b",
    },
    "q37": Object {
      "correct": false,
      "value": "c",
    },
    "q38": Object {
      "correct": false,
      "value": "d",
    },
    "q39": Object {
      "correct": true,
      "value": "a",
    },
    "q4": Object {
      "correct": true,
      "value": "a",
    },
    "q40": Object {
      "correct": false,
      "value": "b",
    },
    "q41": Object {
      "correct": false,
      "value": "c",
    },
    "q42": Object {
      "correct": false,
      "value": "a",
    },
    "q43": Object {
      "correct": true,
      "value": "a",
    },
    "q44": Object {
      "correct": true,
      "value": "c",
    },
    "q45": Object {
      "correct": true,
      "value": "b",
    },
    "q46": Object {
      "correct": false,
      "value": "d",
    },
    "q47": Object {
      "correct": false,
      "value": "a",
    },
    "q48": Object {
      "correct": false,
      "value": "c",
    },
    "q49": Object {
      "correct": true,
      "value": "a",
    },
    "q5": Object {
      "correct": true,
      "value": "c",
    },
    "q50": Object {
      "correct": true,
      "value": "d",
    },
    "q51": Object {
      "correct": false,
      "value": "d",
    },
    "q52": Object {
      "correct": true,
      "value": "b",
    },
    "q53": Object {
      "correct": true,
      "value": "b",
    },
    "q54": Object {
      "correct": true,
      "value": "d",
    },
    "q55": Object {
      "correct": false,
      "value": "d",
    },
    "q56": Object {
      "correct": false,
      "value": "d",
    },
    "q57": Object {
      "correct": false,
      "value": "b",
    },
    "q58": Object {
      "correct": false,
      "value": "b",
    },
    "q59": Object {
      "correct": false,
      "value": "d",
    },
    "q6": Object {
      "correct": true,
      "value": "d",
    },
    "q60": Object {
      "correct": false,
      "value": "d",
    },
    "q7": Object {
      "correct": true,
      "value": "c",
    },
    "q8": Object {
      "correct": true,
      "value": "a",
    },
    "q9": Object {
      "correct": false,
      "value": "c",
    },
  },
  "correctCount": 21,
  "filePath": Any<String>,
  "id": Any<String>,
  "incorrectCount": 39,
  "isCompiled": true,
  "isRollNoExtracted": false,
  "marks": 24,
  "post": 1,
  "questionPaperType": 1,
  "rollNo": undefined,
  "skippedCount": 0,
  "testCenter": 1,
  "testTime": 1,
  "totalMarks": 180,
  "unattemptedCount": 1,
}
`;

exports[`workerCompile should compile just fine 2`] = `
Object {
  "answers": Object {
    "q1": Object {
      "correct": true,
      "value": "d",
    },
    "q10": Object {
      "correct": false,
      "value": "d",
    },
    "q11": Object {
      "correct": false,
      "value": "c",
    },
    "q12": Object {
      "correct": true,
      "value": "d",
    },
    "q13": Object {
      "correct": false,
      "value": "c",
    },
    "q14": Object {
      "correct": false,
      "value": "b",
    },
    "q15": Object {
      "correct": false,
      "value": "a",
    },
    "q16": Object {
      "correct": false,
      "value": "b",
    },
    "q17": Object {
      "correct": false,
      "value": "c",
    },
    "q18": Object {
      "correct": false,
      "value": "b",
    },
    "q19": Object {
      "correct": true,
      "value": "c",
    },
    "q2": Object {
      "correct": false,
      "value": "d",
    },
    "q20": Object {
      "correct": false,
      "value": "a",
    },
    "q21": Object {
      "correct": false,
      "value": "d",
    },
    "q22": Object {
      "correct": false,
      "value": "d",
    },
    "q23": Object {
      "correct": false,
      "unattempted": true,
      "value": "?",
    },
    "q24": Object {
      "correct": false,
      "value": "d",
    },
    "q25": Object {
      "correct": false,
      "value": "d",
    },
    "q26": Object {
      "correct": false,
      "value": "c",
    },
    "q27": Object {
      "correct": false,
      "value": "c",
    },
    "q28": Object {
      "correct": false,
      "value": "a",
    },
    "q29": Object {
      "correct": false,
      "value": "b",
    },
    "q3": Object {
      "correct": false,
      "value": "c",
    },
    "q30": Object {
      "correct": false,
      "value": "c",
    },
    "q31": Object {
      "correct": true,
      "value": "b",
    },
    "q32": Object {
      "correct": true,
      "value": "a",
    },
    "q33": Object {
      "correct": false,
      "value": "d",
    },
    "q34": Object {
      "correct": true,
      "value": "b",
    },
    "q35": Object {
      "correct": true,
      "value": "a",
    },
    "q36": Object {
      "correct": false,
      "value": "b",
    },
    "q37": Object {
      "correct": false,
      "value": "c",
    },
    "q38": Object {
      "correct": false,
      "value": "d",
    },
    "q39": Object {
      "correct": true,
      "value": "a",
    },
    "q4": Object {
      "correct": true,
      "value": "a",
    },
    "q40": Object {
      "correct": false,
      "value": "b",
    },
    "q41": Object {
      "correct": false,
      "value": "c",
    },
    "q42": Object {
      "correct": false,
      "value": "a",
    },
    "q43": Object {
      "correct": true,
      "value": "a",
    },
    "q44": Object {
      "correct": true,
      "value": "c",
    },
    "q45": Object {
      "correct": true,
      "value": "b",
    },
    "q46": Object {
      "correct": false,
      "value": "d",
    },
    "q47": Object {
      "correct": false,
      "value": "a",
    },
    "q48": Object {
      "correct": false,
      "value": "c",
    },
    "q49": Object {
      "correct": true,
      "value": "a",
    },
    "q5": Object {
      "correct": true,
      "value": "c",
    },
    "q50": Object {
      "correct": true,
      "value": "d",
    },
    "q51": Object {
      "correct": false,
      "value": "d",
    },
    "q52": Object {
      "correct": true,
      "value": "b",
    },
    "q53": Object {
      "correct": true,
      "value": "b",
    },
    "q54": Object {
      "correct": true,
      "value": "d",
    },
    "q55": Object {
      "correct": false,
      "value": "d",
    },
    "q56": Object {
      "correct": false,
      "value": "d",
    },
    "q57": Object {
      "correct": false,
      "value": "b",
    },
    "q58": Object {
      "correct": false,
      "value": "b",
    },
    "q59": Object {
      "correct": false,
      "value": "d",
    },
    "q6": Object {
      "correct": true,
      "value": "d",
    },
    "q60": Object {
      "correct": false,
      "value": "d",
    },
    "q7": Object {
      "correct": true,
      "value": "c",
    },
    "q8": Object {
      "correct": true,
      "value": "a",
    },
    "q9": Object {
      "correct": false,
      "value": "c",
    },
  },
  "correctCount": 21,
  "filePath": Any<String>,
  "id": Any<String>,
  "incorrectCount": 39,
  "isCompiled": true,
  "isRollNoExtracted": true,
  "marks": 24,
  "post": 1,
  "questionPaperType": 1,
  "rollNo": "10023",
  "skippedCount": 0,
  "testCenter": 1,
  "testTime": 1,
  "totalMarks": 180,
  "unattemptedCount": 1,
}
`;

exports[`workerCompile should compile just fine 3`] = `
Object {
  "answers": Object {
    "q1": Object {
      "correct": true,
      "value": "d",
    },
    "q10": Object {
      "correct": false,
      "unattempted": true,
      "value": "?",
    },
    "q11": Object {
      "correct": true,
      "value": "b",
    },
    "q12": Object {
      "correct": true,
      "value": "d",
    },
    "q13": Object {
      "correct": true,
      "value": "b",
    },
    "q14": Object {
      "correct": true,
      "value": "a",
    },
    "q15": Object {
      "correct": true,
      "value": "d",
    },
    "q16": Object {
      "correct": true,
      "value": "c",
    },
    "q17": Object {
      "correct": true,
      "value": "b",
    },
    "q18": Object {
      "correct": true,
      "value": "a",
    },
    "q19": Object {
      "correct": true,
      "value": "c",
    },
    "q2": Object {
      "correct": true,
      "value": "b",
    },
    "q20": Object {
      "correct": true,
      "value": "b",
    },
    "q21": Object {
      "correct": true,
      "value": "a",
    },
    "q22": Object {
      "correct": false,
      "value": "*",
    },
    "q23": Object {
      "correct": true,
      "value": "a",
    },
    "q24": Object {
      "correct": false,
      "unattempted": true,
      "value": "?",
    },
    "q25": Object {
      "correct": true,
      "value": "a",
    },
    "q26": Object {
      "correct": true,
      "value": "b",
    },
    "q27": Object {
      "correct": true,
      "value": "a",
    },
    "q28": Object {
      "correct": true,
      "value": "b",
    },
    "q29": Object {
      "correct": true,
      "value": "d",
    },
    "q3": Object {
      "correct": true,
      "value": "d",
    },
    "q30": Object {
      "correct": true,
      "value": "a",
    },
    "q31": Object {
      "correct": true,
      "value": "b",
    },
    "q32": Object {
      "correct": true,
      "value": "a",
    },
    "q33": Object {
      "correct": true,
      "value": "b",
    },
    "q34": Object {
      "correct": true,
      "value": "b",
    },
    "q35": Object {
      "correct": true,
      "value": "a",
    },
    "q36": Object {
      "correct": true,
      "value": "c",
    },
    "q37": Object {
      "correct": false,
      "unattempted": true,
      "value": "?",
    },
    "q38": Object {
      "correct": true,
      "value": "a",
    },
    "q39": Object {
      "correct": true,
      "value": "a",
    },
    "q4": Object {
      "correct": true,
      "value": "a",
    },
    "q40": Object {
      "correct": true,
      "value": "a",
    },
    "q41": Object {
      "correct": true,
      "value": "a",
    },
    "q42": Object {
      "correct": false,
      "unattempted": true,
      "value": "?",
    },
    "q43": Object {
      "correct": true,
      "value": "a",
    },
    "q44": Object {
      "correct": true,
      "value": "c",
    },
    "q45": Object {
      "correct": true,
      "value": "b",
    },
    "q46": Object {
      "correct": true,
      "value": "a",
    },
    "q47": Object {
      "correct": false,
      "unattempted": true,
      "value": "?",
    },
    "q48": Object {
      "correct": true,
      "value": "b",
    },
    "q49": Object {
      "correct": true,
      "value": "a",
    },
    "q5": Object {
      "correct": true,
      "value": "c",
    },
    "q50": Object {
      "correct": true,
      "value": "d",
    },
    "q51": Object {
      "correct": true,
      "value": "a",
    },
    "q52": Object {
      "correct": true,
      "value": "b",
    },
    "q53": Object {
      "correct": true,
      "value": "b",
    },
    "q54": Object {
      "correct": true,
      "value": "d",
    },
    "q55": Object {
      "correct": false,
      "value": "*",
    },
    "q56": Object {
      "correct": true,
      "value": "b",
    },
    "q57": Object {
      "correct": true,
      "value": "a",
    },
    "q58": Object {
      "correct": true,
      "value": "d",
    },
    "q59": Object {
      "correct": true,
      "value": "a",
    },
    "q6": Object {
      "correct": true,
      "value": "d",
    },
    "q60": Object {
      "correct": true,
      "value": "a",
    },
    "q7": Object {
      "correct": true,
      "value": "c",
    },
    "q8": Object {
      "correct": true,
      "value": "a",
    },
    "q9": Object {
      "correct": true,
      "value": "d",
    },
  },
  "correctCount": 53,
  "filePath": Any<String>,
  "id": Any<String>,
  "incorrectCount": 7,
  "isCompiled": true,
  "isRollNoExtracted": true,
  "marks": 152,
  "post": 1,
  "questionPaperType": 1,
  "rollNo": "10025",
  "skippedCount": 0,
  "testCenter": 1,
  "testTime": 1,
  "totalMarks": 180,
  "unattemptedCount": 5,
}
`;
